### PR TITLE
Add MetricsContainer::getPerWorker{Counter|Histogram} (#28903)

### PR DESCRIPTION
**Pull request by JayajP**
Add getPerWorker{Counter|Histogram} methods to the MetricsContainer interface. By default, these methods will return no-op metrics that do not record any values.

For the Legacy Java runner, perWorkerMetrics are stored when the experiment `enable_per_worker_metrics` is enabled. Otherwise, the Legacy runner defaults to the no-op metrics. 

These methods will be used to to store BigQuery related metrics that will be piped back to WMW.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/NoOpCounter.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.metrics;
+
+/** A no-op implementation of Counter. */
+public class NoOpCounter implements Counter {
```
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/NoOpCounter.java:21]: Can you add some commentary here to describe why this exists? Just a quick comment to describe that it exists in order to provide a default if the implementation doesn't override it?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 2]
```diff
@@ -463,6 +463,11 @@ public static void main(String[] args) throws Exception {
     // metrics.
     MetricsEnvironment.setProcessWideContainer(new MetricsLogger(null));
 
+    // When enabled, the Pipeline will record Per-Worker metrics that will be piped to WMW.
+    StreamingStepMetricsContainer.setEnablePerWorkerMetrics(
+        options.isEnableStreamingEngine()
+            && DataflowRunner.hasExperiment(options, "enable_per_worker_metrics"));
```
Comment (johnjcasey) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:469]: Should this be an experiment, or just on by default?

**Comment by JayajP** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/NoOpCounter.java (hunk 1 above)
Comment (JayajP) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/NoOpCounter.java:24]: done

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 2 above)
Comment (JayajP) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:469]: Will leave this as an experiment until I've done some load tests. Then I will enable it as default for Streaming Engine Jobs.

**Review by johnjcasey — APPROVED** [review]

**Comment by svetakvsundhar** [comment]
I think the precommit Java GCP IO Direct failure might be related to the GHA migration, and so it may not actually be an issue.

cc: @Abacn to confirm

**Pull request merged**